File: Python3/44. Wildcard Matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        # s index, p index
        si = pi = 0

        # s length
        sl = len(s)

        # p length
        pl = len(p)

        # * index, temp * index
        star = temp = -1

        while si < sl:
            # match pattern: same char or '?'
            if pi < pl and p[pi] in ('?', s[si]):
                si += 1
                pi += 1
            
            # match pattern: '*'
            elif pi < pl and p[pi] == '*':
                star = pi
                temp = si
                pi += 1

            # no more patterns or '*'
            elif star == -1:
                return False

            #no more patterns but has star
            else:
                pi = star + 1
                si = temp + 1
                temp = si

        # remaining chars in p should all be '*'
        return all(p[i] == '*' for i in range(pi, pl))

File: Java/44. Wildcard Matching.java
class Solution {
    public boolean isMatch(String s, String p) {
        // s index, p index
        int si = 0, pi = 0;

        // s length, p length
        int sl = s.length(), pl = p.length();

        // star index, temp index
        int star = -1, temp = -1;

        while (si < sl) {
            if (pi < pl && (p.charAt(pi) == s.charAt(si) || p.charAt(pi) == '?')) {
                si++;
                pi++;
            } else if (pi < pl && p.charAt(pi) == '*') {
                star = pi;
                temp = si;
                pi++;
            } else if (star == -1) {
                return false;
            } else {
                pi = star + 1;
                si = temp + 1;
                temp = si;
            }
        }

        for (int i = pi; i < pl; i++) {
            if (p.charAt(i) != '*') {
                return false;
            }
        }

        return true;
    }
}